spec/classes/collectd_plugin_dpdk_telemetry_spec.rb

```diff
@@ -0,0 +1,67 @@
+require 'spec_helper'
+
+describe 'collectd::plugin::dpdk_telemetry', type: :class do
+  on_supported_os(baseline_os_hash).each do |os, facts|
+    context "on #{os} " do
+      let :facts do
+        facts
+      end
+
+      options = os_specific_options(facts)
+
+      context ':ensure => present, default params' do
+        content = <<EOS
+# Generated by Puppet
+<LoadPlugin dpdk_telemetry>
+  Globals false
+</LoadPlugin>
+
+<Plugin dpdk_telemetry>
+  ClientSocketPath "/var/run/.client"
+  DpdkSocketPath "/var/run/dpdk/rte/telemetry"
+</Plugin>
+
+EOS
+
+        it "Will create #{options[:plugin_conf_dir]}/10-dpdk_telemetry.conf" do
+          is_expected.to compile.with_all_deps
+          is_expected.to contain_file('dpdk_telemetry.load').with(
+            ensure: 'present',
+            path: "#{options[:plugin_conf_dir]}/10-dpdk_telemetry.conf",
+            content: content
+          )
+        end
+      end
+
+      context ':ensure => present and :client_socket_path => /test/path/.client' do
+        let :params do
+          { client_socket_path: '/test/path/.client' }
+        end
+
+        it "Will create #{options[:plugin_conf_dir]}/10-dpdk_telemetry.conf" do
+          is_expected.to compile.with_all_deps
+          is_expected.to contain_file('dpdk_telemetry.load').with(
+            ensure: 'present',
+            path: "#{options[:plugin_conf_dir]}/10-dpdk_telemetry.conf",
+            content: %r{ClientSocketPath "/test/path/.client"}m
+          )
+        end
+      end
+
+      context ':ensure => present and :dpdk_socket_path => /test/path/telemetry' do
+        let :params do
+          { dpdk_socket_path: '/test/path/telemetry' }
+        end
+
+        it "Will create #{options[:plugin_conf_dir]}/10-dpdk_telemetry.conf" do
```

Comment:
Since `client_socket_path` can be `undef`, there should be a test that shows the content when the param is `undef`. Same for `dpdk_socket_path`.

Reply:
I was looking into creating test case that you described but I stumbled upon obstacle. I wrote following test:

```
      context ':ensure => present, :client_socket_path => undef and :dpdk_socket_path => undef' do
        let :params do
          { client_socket_path: :undef,
            dpdk_socket_path: :undef }
        end

        sth = <<EOS
# Generated by Puppet
<LoadPlugin dpdk_telemetry>
  Globals false
</LoadPlugin>

<Plugin dpdk_telemetry>
</Plugin>

EOS

        it "Will create #{options[:plugin_conf_dir]}/10-dpdk_telemetry.conf" do
          is_expected.to compile.with_all_deps
          is_expected.to contain_file('dpdk_telemetry.load').with(
            ensure: 'present',
            path: "#{options[:plugin_conf_dir]}/10-dpdk_telemetry.conf",
            content: sth
          )
        end
      end
```
	 
Configuration is generated using specified default values rather than assuming that no value was provided.
I found interesting message by _jcbollinger_ [here](https://groups.google.com/d/msg/puppet-users/6nL2eROH8is/Fg6W5lEjHQsJ), suggesting that assigning undef doesn't actually assign undef, only implies that no specified value is being assigned, and therefore default should be used.

I'm relatively new to Puppet, RSpec and Ruby so I can't be sure, however my observations suggests that it in fact might work that way. Additionally, I couldn't find this kind of test case in this repository. Are you sure that test case you described is possible to write? If so, could you give any suggestion how to proceed?